libs/custom-paper-handler/src/driver/driver.ts

```diff
@@ -477,53 +443,95 @@ export class PaperHandlerDriver {
     await this.transferOutGeneric(ScanCommand, undefined);
     debug('STARTING SCAN');
     let scanStatus = OK_CONTINUE;
+    let dataBlockBytesReceived = 0;
+    let width = -1;
     const imageData: Uint8Array[] = [];
 
+    // Assumptions:
+    // 1. There's at least 1 data block
+    // 2. Each data block can fit within a single `transferInGeneric` call
     while (scanStatus === OK_CONTINUE) {
       const rawResponse = await this.transferInGeneric();
       assert(rawResponse?.data);
 
       const responseBuffer = rawResponse.data.buffer;
-
       const header = responseBuffer.slice(0, SCAN_HEADER_LENGTH_BYTES);
       const response: ScanResponse = ScanResponse.decode(
         Buffer.from(header)
       ).unsafeUnwrap();
-
       scanStatus = response.returnCode;
       const { sizeX, sizeY } = response;
-      const pixelsPerByte = response.scan <= 5 ? 1 : 8;
+      // The last sizeX returned by the scan command is 0, so store the first one for use later
+      if (width === -1) {
+        width = sizeX;
+      }
+      const pixelsPerByte = response.scan <= 0x05 ? 1 : 8;
       debug(`sizeX: ${sizeX}, sizeY: ${sizeY}, ppb: ${pixelsPerByte}`);
-
       const dataBlockByteLength = (sizeX * sizeY) / pixelsPerByte;
-      imageData.push(
-        new Uint8Array(responseBuffer.slice(SCAN_HEADER_LENGTH_BYTES))
-      );
-
-      let dataBlockBytesReceived = responseBuffer.byteLength - 16;
+      const dataBlock = responseBuffer.slice(SCAN_HEADER_LENGTH_BYTES);
+      dataBlockBytesReceived += dataBlock.byteLength;
       debug(
-        `initial ${dataBlockBytesReceived} bytes received. Expecting ${dataBlockByteLength}`
+        `Expected ${dataBlockByteLength} bytes, got ${dataBlock.byteLength} in this block, ${dataBlockBytesReceived} so far`
       );
-
-      while (dataBlockBytesReceived < dataBlockByteLength) {
-        debug(
-          `${dataBlockBytesReceived} bytes received. Expecting ${dataBlockByteLength}`
-        );
-        await sleep(1000);
-        debug('Additional data...');
-        const { data } = await this.transferInGeneric();
-        assert(data);
-        debug(`${data.byteLength}`);
-        imageData.push(new Uint8Array(data.buffer));
-        dataBlockBytesReceived += data.byteLength;
-      }
+      imageData.push(new Uint8Array(dataBlock));
     }
+
     this.genericLock.release();
     debug('ALL BLOCKS RECEIVED');
+    if (scanStatus !== OK_NO_MORE_DATA) {
+      throw new Error(
+        `Unhandled scan result status: ${scanStatus
+          .toString(16)
+          .padStart(2, '0')
+          .toUpperCase()}`
+      );
+    }
+    const imageBuf = Buffer.concat(imageData);
     return createImageData(
-      Uint8ClampedArray.from(Buffer.concat(imageData)),
-      1728
+      Uint8ClampedArray.from(imageBuf),
+      // Start scan ticket (0x1c 0x53 0x50 0x53) returns width of data block [Size X]; hardcode this value for now but
+      // we should read it from the response header
```

Comment:
What's the hardcoded part?

Reply:
Outdated comment, sorry for leaving that in. We are actually reading width from the response header.